### PR TITLE
Fix/queue induction email

### PR DESCRIPTION
**Pull request by nishant22029**
- Queue induction email  when contact submits volunteer form after contribution email field is not saved .

## Summary by CodeRabbit

- **New Features**
	- Introduced a queuing system for induction emails, enhancing email processing efficiency.
	- Added functionality to check if an induction email has already been sent to prevent duplicates.
	- Improved retrieval of the volunteer induction coordinator's email for better communication.
	- Added a method to process queued induction emails, ensuring reliable sending of notifications.

- **Bug Fixes**
	- Enhanced error handling for email sending logic to manage exceptions effectively.

### CODE REVIEW

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php [hunk 1]
```diff
@@ -187,16 +189,89 @@ private static function sendInductionEmail($contactId) {
...
+   * Process the queued induction email task.
+   */
+  public static function processQueuedInductionEmail($queue, $params) {
+    try {
+      $result = civicrm_api3('Email', 'send', $params);
+      if ($result['is_error']) {
+        throw new \CRM_Core_Exception($result['error_message']);
+      }
+      \Civi::log()->info('Successfully sent queued induction email', [
+        'contactId' => $params['contact_id'],
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:261]: can we log the `$params` as well?

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php [hunk 2]
```diff
@@ -187,16 +189,89 @@ private static function sendInductionEmail($contactId) {
...
+    }
+    catch (\CRM_Core_Exception $ex) {
+      \Civi::log()->error('Failed to queue induction email due to CiviCRM error', [
+        'contactId' => $params['contact_id'],
+        'error' => $ex->getMessage(),
+      ]);
+    }
+    catch (\Exception $ex) {
+      \Civi::log()->error('Unexpected error while queueing induction email', [
+        'contactId' => $params['contact_id'],
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:245]: let's log all the params and not just contactId
```suggestion
        'paramas' => $params,
```

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php [hunk 3]
```diff
@@ -187,16 +189,89 @@ private static function sendInductionEmail($contactId) {
...
+        ]);
+
+      \Civi::log()->info('Induction email queued for contact', ['contactId' => $params['contact_id']]);
+    }
+    catch (\CRM_Core_Exception $ex) {
+      \Civi::log()->error('Failed to queue induction email due to CiviCRM error', [
+        'contactId' => $params['contact_id'],
+        'error' => $ex->getMessage(),
+      ]);
+    }
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:242]: Why do we have 2 catch blocks?

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php [hunk 4]
```diff
@@ -173,11 +173,13 @@ private static function findCoordinatorForOffice($officeId) {
    * Common logic to send an email.
    */
   private static function sendInductionEmail($contactId) {
```
Comment (pokhiii) [inline wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:175]: can we say `volunteerId` instead of `contactId`?

**Review by pokhiii — CHANGES_REQUESTED** [review]

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php (hunk 1 above)
Comment (nishant22029) [reply wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:261]: sure

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php (hunk 4 above)
Comment (nishant22029) [reply wp-content/civi-extensions/goonjcustom/Civi/InductionService.php]: SURE

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php (hunk 3 above)
Comment (nishant22029) [reply wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:242]: removed one

**Review by pokhiii — CHANGES_REQUESTED** [review]

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php [hunk 5]
```diff
@@ -172,12 +172,14 @@ private static function findCoordinatorForOffice($officeId) {
   /**
    * Common logic to send an email.
    */
-  private static function sendInductionEmail($contactId) {
-    if (self::emailAlreadySent($contactId)) {
-      \Civi::log()->info('Induction email already sent for contact', ['id' => $contactId]);
+  private static function sendInductionEmail($volunteerId) {
+    // Check if the email was already sent.
+    if (self::isEmailAlreadySent($contactId)) {
+      \Civi::log()->info('Induction email already sent for contact', ['id' => $volunteerId]);
```
[the remark this answers is not in the document]
Comment (nishant22029) [reply wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:178]: sure removed

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php [hunk 6]
```diff
@@ -187,16 +189,83 @@ private static function sendInductionEmail($contactId) {
...
+    // Prepare email parameters.
     $emailParams = [
-      'contact_id' => $contactId,
+      'contact_id' => $volunteerId,
       'template_id' => $template['id'],
       'cc' => self::$volunteerInductionAssigneeEmail,
     ];
 
+    // Fetch contact's email.
+    $contacts = Contact::get(FALSE)
```
[the remark this answers is not in the document]
Comment (nishant22029) [reply wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:200]: removed this now

File: wp-content/civi-extensions/goonjcustom/Civi/InductionService.php [hunk 7]
```diff
@@ -172,12 +172,14 @@ private static function findCoordinatorForOffice($officeId) {
   /**
    * Common logic to send an email.
    */
-  private static function sendInductionEmail($contactId) {
-    if (self::emailAlreadySent($contactId)) {
-      \Civi::log()->info('Induction email already sent for contact', ['id' => $contactId]);
+  private static function sendInductionEmail($volunteerId) {
+    // Check if the email was already sent.
```
[the remark this answers is not in the document]
Comment (nishant22029) [reply wp-content/civi-extensions/goonjcustom/Civi/InductionService.php:176]: sure

**Review by pokhiii — APPROVED** [review]

**Pull request merged**